File: backend/routes/phase4_routes.py
"""Phase 4a routes — reviews, notifications, wallet, payments, subs, KYC, trust, admin/kyc."""
from __future__ import annotations
from fastapi import APIRouter, Depends, HTTPException, Query, Request
from pydantic import BaseModel, Field
from typing import Literal

from middleware.auth_middleware import require_auth
from services import (
    phase4_service as p4, wallet_service, notification_service, razorpay_service,
)

router = APIRouter(prefix="/v1", tags=["phase4"])


def _require_admin(u):
    if "admin" not in (u.roles or []):
        raise HTTPException(403, "Admin only")


# =============== Reviews ===============
class ReviewBody(BaseModel):
    target_type: Literal["vendor", "listing", "service"]
    target_id: str
    rating: int
    comment: str | None = None
    images: list[dict] = Field(default_factory=list)
    videos: list[dict] = Field(default_factory=list)
    deal_id: str | None = None
    listing_id: str | None = None


class ReviewUpdate(BaseModel):
    rating: int | None = None
    comment: str | None = None
    images: list[dict] | None = None
    videos: list[dict] | None = None


class ReplyBody(BaseModel):
    text: str = Field(..., min_length=1, max_length=500)


@router.post("/reviews/")
async def create_review(body: ReviewBody, user=Depends(require_auth)):
    return await p4.create_review(user.id, body.model_dump(exclude_none=True))


@router.get("/reviews/")
async def list_reviews(target_type: str, target_id: str, sort: str = "recent",
                      cursor: str | None = None, limit: int = Query(20, ge=1, le=50)):
    return await p4.list_reviews(target_type, target_id, sort, cursor, limit)


@router.patch("/reviews/{rid}")
async def update_review(rid: str, body: ReviewUpdate, user=Depends(require_auth)):
    from bson import ObjectId
    if not ObjectId.is_valid(rid):
        raise HTTPException(400, "Invalid id")
    return await p4.update_review(rid, user.id, body.model_dump(exclude_none=True))


@router.delete("/reviews/{rid}")
async def delete_review(rid: str, user=Depends(require_auth)):
    from bson import ObjectId
    if not ObjectId.is_valid(rid):
        raise HTTPException(400, "Invalid id")
    await p4.soft_delete_review(rid, user.id, is_admin="admin" in (user.roles or []))
    return {"ok": True}


@router.post("/reviews/{rid}/reply")
async def reply_review(rid: str, body: ReplyBody, user=Depends(require_auth)):
    from bson import ObjectId
    if not ObjectId.is_valid(rid):
        raise HTTPException(400, "Invalid id")
    return await p4.reply_to_review(rid, user.id, body.text)


@router.post("/reviews/{rid}/helpful")
async def helpful_review(rid: str, user=Depends(require_auth)):
    return await p4.toggle_helpful(rid, user.id)


@router.get("/reviews/vendor/{vendor_id}/summary")
async def vendor_review_summary(vendor_id: str):
    return await p4.summary("vendor", vendor_id)


# =============== Notifications ===============
@router.get("/notifications/me")
async def my_notifs(is_read: bool | None = None, cursor: str | None = None,
                    limit: int = Query(30, ge=1, le=100), user=Depends(require_auth)):
    return await notification_service.list_mine(user.id, is_read, cursor, limit)


@router.get("/notifications/me/unread-count")
async def notif_unread(user=Depends(require_auth)):
    return {"count": await notification_service.unread_count(user.id)}


@router.post("/notifications/{nid}/read")
async def read_notif(nid: str, user=Depends(require_auth)):
    await notification_service.mark_read(nid, user.id)
    return {"ok": True}


@router.post("/notifications/me/read-all")
async def read_all(user=Depends(require_auth)):
    await notification_service.mark_all_read(user.id)
    return {"ok": True}


@router.delete("/notifications/{nid}")
async def dismiss(nid: str, user=Depends(require_auth)):
    await notification_service.dismiss(nid, user.id)
    return {"ok": True}


# =============== Wallet ===============
@router.get("/wallet/me")
async def wallet_me(user=Depends(require_auth)):
    w = await wallet_service.get_or_create(user.id)
    txns = await wallet_service.list_transactions(user.id, 20)
    return {"wallet": w, "transactions": txns}


@router.get("/wallet/me/transactions")
async def wallet_txns(limit: int = Query(50, ge=1, le=200), user=Depends(require_auth)):
    return {"items": await wallet_service.list_transactions(user.id, limit)}


class TopupBody(BaseModel):
    amount_paise: int = Field(..., ge=100)


@router.post("/wallet/me/topup")
async def wallet_topup(body: TopupBody, user=Depends(require_auth)):
    return await p4.create_payment_order(user.id, "wallet_topup", body.amount_paise)


# =============== Payments ===============
class PaymentOrderBody(BaseModel):
    purpose: Literal["verified_badge_monthly", "verified_badge_yearly", "wallet_topup", "listing_boost", "other"]
    amount_paise: int | None = None
    ref_id: str | None = None


class VerifyBody(BaseModel):
    razorpay_order_id: str
    razorpay_payment_id: str
    razorpay_signature: str


@router.post("/payments/order")
async def create_order(body: PaymentOrderBody, user=Depends(require_auth)):
    amt = body.amount_paise
    if body.purpose.startswith("verified_badge"):
        amt = p4.PLAN_PAISE.get(f"verified_{body.purpose.split('_')[-1]}") or amt
    if not amt or amt <= 0:
        raise HTTPException(400, "amount required")
    return await p4.create_payment_order(user.id, body.purpose, amt, body.ref_id)


@router.post("/payments/verify")
async def verify_payment(body: VerifyBody, user=Depends(require_auth)):
    return await p4.verify_and_capture(user.id, body.razorpay_order_id, body.razorpay_payment_id, body.razorpay_signature)


@router.post("/payments/dev/simulate-success")
async def simulate_success(payment_id: str, user=Depends(require_auth)):
    return await p4.dev_simulate_success(payment_id, user.id)


@router.get("/payments/me")
async def my_payments(user=Depends(require_auth)):
    return {"items": await p4.my_payments(user.id)}


@router.post("/payments/webhook")
async def razorpay_webhook(request: Request):
    body = await request.body()
    sig = request.headers.get("x-razorpay-signature", "")
    if not razorpay_service.verify_webhook_signature(body, sig):
        raise HTTPException(400, "Invalid webhook signature")
    import json
    try:
        payload = json.loads(body)
    except Exception:  # noqa: BLE001
        raise HTTPException(400, "Bad payload")
    event = payload.get("event", "")
    pay_entity = (payload.get("payload", {}).get("payment") or {}).get("entity") or {}
    order_id = pay_entity.get("order_id")
    if event == "payment.captured" and order_id:
        from database import get_db
        db = get_db()
        pdoc = await db.payments.find_one({"razorpay_order_id": order_id})
        if pdoc and pdoc.get("status") != "captured":
            await p4._apply_success(pdoc, pay_entity.get("id", ""), request.headers.get("x-razorpay-signature", ""))
    return {"ok": True}


# =============== Subscriptions ===============
class SubBody(BaseModel):
    plan: Literal["verified_monthly", "verified_yearly"]


@router.post("/subscriptions/subscribe")
async def subscribe(body: SubBody, user=Depends(require_auth)):
    return await p4.create_sub_order(user.id, body.plan)


@router.get("/subscriptions/me")
async def my_subs(user=Depends(require_auth)):
    return {"items": await p4.my_subs(user.id)}


@router.post("/subscriptions/{sid}/cancel")
async def cancel_sub(sid: str, user=Depends(require_auth)):
    return await p4.cancel_sub(sid, user.id)


# =============== KYC ===============
class KycBody(BaseModel):
    doc_type: Literal["aadhaar", "pan", "driving_license", "passport"]
    doc_number: str = Field(..., min_length=4, max_length=32)
    doc_url: str
    selfie_url: str | None = None


@router.post("/kyc/me/submit")
async def kyc_submit(body: KycBody, user=Depends(require_auth)):
    return await p4.kyc_submit(user.id, body.model_dump(exclude_none=True))


@router.get("/kyc/me")
async def kyc_me(user=Depends(require_auth)):
    return await p4.my_kyc(user.id) or {"status": "unverified"}


# =============== Admin ===============
class KycActionBody(BaseModel):
    reason: str | None = None


@router.get("/admin/kyc")
async def admin_kyc_queue(user=Depends(require_auth)):
    _require_admin(user)
    return {"items": await p4.kyc_queue()}


@router.post("/admin/kyc/{kid}/approve")
async def admin_approve(kid: str, user=Depends(require_auth)):
    _require_admin(user)
    return await p4.kyc_review(kid, user.id, approve=True)


@router.post("/admin/kyc/{kid}/reject")
async def admin_reject(
    kid: str,
    body: KycActionBody | None = None,
    reason: str | None = None,  # query param for backward compat
    user=Depends(require_auth),
):
    _require_admin(user)
    final_reason = (body.reason if body else None) or reason
    return await p4.kyc_review(kid, user.id, approve=False, reason=final_reason)


# =============== Trust Score ===============
@router.get("/users/{user_id}/trust-score")
async def trust(user_id: str):
    return await p4.trust_score(user_id)

File: backend/src/routes/phase4.routes.js
const express = require('express');
const { requireAuth } = require('../middleware/auth.middleware');
const reviewService = require('../services/review.service');
const walletService = require('../services/wallet.service');
const notificationService = require('../services/notification.service');
const paymentService = require('../services/payment.service');
const razorpayService = require('../services/razorpay.service');
const subscriptionService = require('../services/subscription.service');
const kycService = require('../services/kyc.service');
const trustService = require('../services/trust.service');
const Payment = require('../models/Phase4').Payment;
const { catchAsync } = require('../utils/helpers');
const ApiError = require('../utils/ApiError');

const router = express.Router();

const requireAdmin = (req, res, next) => {
  const roles = req.user.roles || [];
  if (!roles.includes('admin')) {
    return next(ApiError.forbidden('Admin only'));
  }
  next();
};

const PLAN_PAISE = {
  verified_monthly: 29900,
  verified_yearly: 299900,
};

// ============================================================ REVIEWS
router.post('/reviews', requireAuth, catchAsync(async (req, res) => {
  const result = await reviewService.createReview(req.user._id.toString(), req.body);
  res.json(result);
}));

router.get('/reviews', catchAsync(async (req, res) => {
  const { target_type, target_id, sort = 'recent', cursor } = req.query;
  const limit = Math.max(1, Math.min(50, parseInt(req.query.limit || 20, 10)));
  if (!target_type || !target_id) {
    throw ApiError.badRequest('target_type and target_id are required');
  }

  const result = await reviewService.listReviews(target_type, target_id, sort, cursor, limit);
  res.json(result);
}));

router.patch('/reviews/:rid', requireAuth, catchAsync(async (req, res) => {
  const result = await reviewService.updateReview(req.params.rid, req.user._id.toString(), req.body);
  res.json(result);
}));

router.delete('/reviews/:rid', requireAuth, catchAsync(async (req, res) => {
  const is_admin = req.user.roles && req.user.roles.includes('admin');
  await reviewService.softDeleteReview(req.params.rid, req.user._id.toString(), is_admin);
  res.json({ ok: true });
}));

router.post('/reviews/:rid/reply', requireAuth, catchAsync(async (req, res) => {
  const { text } = req.body;
  if (!text) {
    throw ApiError.badRequest('text is required');
  }
  const result = await reviewService.replyToReview(req.params.rid, req.user._id.toString(), text);
  res.json(result);
}));

router.post('/reviews/:rid/helpful', requireAuth, catchAsync(async (req, res) => {
  const result = await reviewService.toggleHelpful(req.params.rid, req.user._id.toString());
  res.json(result);
}));

router.get('/reviews/vendor/:vendor_id/summary', catchAsync(async (req, res) => {
  const result = await reviewService.summary('vendor', req.params.vendor_id);
  res.json(result);
}));

// ============================================================ NOTIFICATIONS
router.get('/notifications/me', requireAuth, catchAsync(async (req, res) => {
  const isRead = req.query.is_read !== undefined ? req.query.is_read === 'true' : null;
  const limit = Math.max(1, Math.min(100, parseInt(req.query.limit || 30, 10)));

  const result = await notificationService.listMine(req.user._id.toString(), isRead, req.query.cursor || null, limit);
  res.json(result);
}));

router.get('/notifications/me/unread-count', requireAuth, catchAsync(async (req, res) => {
  const count = await notificationService.unreadCount(req.user._id.toString());
  res.json({ count });
}));

router.post('/notifications/:nid/read', requireAuth, catchAsync(async (req, res) => {
  await notificationService.markRead(req.params.nid, req.user._id.toString());
  res.json({ ok: true });
}));

router.post('/notifications/me/read-all', requireAuth, catchAsync(async (req, res) => {
  await notificationService.markAllRead(req.user._id.toString());
  res.json({ ok: true });
}));

router.delete('/notifications/:nid', requireAuth, catchAsync(async (req, res) => {
  await notificationService.dismiss(req.params.nid, req.user._id.toString());
  res.json({ ok: true });
}));

// ============================================================ WALLET
router.get('/wallet/me', requireAuth, catchAsync(async (req, res) => {
  const w = await walletService.getOrCreate(req.user._id.toString());
  const txns = await walletService.listTransactions(req.user._id.toString(), 20);
  res.json({ wallet: w, transactions: txns });
}));

router.get('/wallet/me/transactions', requireAuth, catchAsync(async (req, res) => {
  const limit = Math.max(1, Math.min(200, parseInt(req.query.limit || 50, 10)));
  const txns = await walletService.listTransactions(req.user._id.toString(), limit);
  res.json({ items: txns });
}));

router.post('/wallet/me/topup', requireAuth, catchAsync(async (req, res) => {
  const { amount_paise } = req.body;
  if (!amount_paise || amount_paise < 100) {
    throw ApiError.badRequest('amount_paise must be at least 100');
  }

  const order = await paymentService.createPaymentOrder(req.user._id.toString(), 'wallet_topup', amount_paise);
  res.json(order);
}));

// ============================================================ PAYMENTS
router.post('/payments/order', requireAuth, catchAsync(async (req, res) => {
  const { purpose, amount_paise, ref_id } = req.body;
  if (!purpose) {
    throw ApiError.badRequest('purpose is required');
  }

  let amt = amount_paise;
  if (purpose.startsWith('verified_badge')) {
    const planSuffix = purpose.split('_').pop(); // monthly / yearly
    amt = PLAN_PAISE[`verified_${planSuffix}`] || amt;
  }

  if (!amt || amt <= 0) {
    throw ApiError.badRequest('amount required');
  }

  const order = await paymentService.createPaymentOrder(req.user._id.toString(), purpose, amt, ref_id || null);
  res.json(order);
}));

router.post('/payments/verify', requireAuth, catchAsync(async (req, res) => {
  const { razorpay_order_id, razorpay_payment_id, razorpay_signature } = req.body;
  if (!razorpay_order_id || !razorpay_payment_id || !razorpay_signature) {
    throw ApiError.badRequest('Missing Razorpay verification tokens');
  }

  const result = await paymentService.verifyAndCapture(
    req.user._id.toString(),
    razorpay_order_id,
    razorpay_payment_id,
    razorpay_signature
  );
  res.json(result);
}));

router.post('/payments/dev/simulate-success', requireAuth, catchAsync(async (req, res) => {
  const { payment_id } = req.body;
  if (!payment_id) {
    throw ApiError.badRequest('payment_id is required');
  }

  const result = await paymentService.devSimulateSuccess(payment_id, req.user._id.toString());
  res.json(result);
}));

router.get('/payments/me', requireAuth, catchAsync(async (req, res) => {
  const items = await paymentService.myPayments(req.user._id.toString());
  res.json({ items });
}));

router.post('/payments/webhook', catchAsync(async (req, res) => {
  const rawBody = req.rawBody; // Assumes raw body parser handles signature
  const sig = req.headers['x-razorpay-signature'] || '';
  if (!rawBody || !razorpayService.verifyWebhookSignature(rawBody, sig)) {
    throw ApiError.badRequest('Invalid webhook signature');
  }

  let payload;
  try {
    payload = JSON.parse(rawBody.toString('utf-8'));
  } catch (err) {
    throw ApiError.badRequest('Bad payload');
  }

  const event = payload.event;
  const payEntity = payload.payload?.payment?.entity || {};
  const orderId = payEntity.order_id;

  if (event === 'payment.captured' && orderId) {
    const pdoc = await Payment.findOne({ razorpay_order_id: orderId });
    if (pdoc && pdoc.status !== 'captured') {
      await paymentService.applySuccess(pdoc, payEntity.id, sig);
    }
  }

  res.json({ ok: true });
}));

// ============================================================ SUBSCRIPTIONS
router.post('/subscriptions/subscribe', requireAuth, catchAsync(async (req, res) => {
  const { plan } = req.body;
  if (!['verified_monthly', 'verified_yearly'].includes(plan)) {
    throw ApiError.badRequest('Invalid plan');
  }

  const result = await subscriptionService.createSubOrder(req.user._id.toString(), plan);
  res.json(result);
}));

router.get('/subscriptions/me', requireAuth, catchAsync(async (req, res) => {
  const items = await subscriptionService.mySubs(req.user._id.toString());
  res.json({ items });
}));

router.post('/subscriptions/:sid/cancel', requireAuth, catchAsync(async (req, res) => {
  const result = await subscriptionService.cancelSub(req.params.sid, req.user._id.toString());
  res.json(result);
}));

// ============================================================ KYC
router.post('/kyc/me/submit', requireAuth, catchAsync(async (req, res) => {
  const { doc_type, doc_number, doc_url, selfie_url } = req.body;
  if (!['aadhaar', 'pan', 'driving_license', 'passport'].includes(doc_type)) {
    throw ApiError.badRequest('Invalid doc_type');
  }
  if (!doc_number || doc_number.length < 4 || doc_number.length > 32) {
    throw ApiError.badRequest('Invalid doc_number');
  }
  if (!doc_url) {
    throw ApiError.badRequest('doc_url is required');
  }

  const result = await kycService.kycSubmit(req.user._id.toString(), {
    doc_type,
    doc_number,
    doc_url,
    selfie_url: selfie_url || null,
  });
  res.json(result);
}));

router.get('/kyc/me', requireAuth, catchAsync(async (req, res) => {
  const result = await kycService.myKyc(req.user._id.toString());
  res.json(result || { status: 'unverified' });
}));

// ============================================================ ADMIN KYC
router.get('/admin/kyc', requireAuth, requireAdmin, catchAsync(async (req, res) => {
  const items = await kycService.kycQueue();
  res.json({ items });
}));

router.post('/admin/kyc/:kid/approve', requireAuth, requireAdmin, catchAsync(async (req, res) => {
  const result = await kycService.kycReview(req.params.kid, req.user._id.toString(), true);
  res.json(result);
}));

router.post('/admin/kyc/:kid/reject', requireAuth, requireAdmin, catchAsync(async (req, res) => {
  const reason = req.body?.reason || req.query.reason || null;
  const result = await kycService.kycReview(req.params.kid, req.user._id.toString(), false, reason);
  res.json(result);
}));

// ============================================================ TRUST SCORE
router.get('/users/:user_id/trust-score', catchAsync(async (req, res) => {
  const result = await trustService.getTrustScore(req.params.user_id);
  res.json(result);
}));

module.exports = router;
